src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingMiddleware.cs

```diff
@@ -32,29 +37,31 @@ public ResponseCachingMiddleware(RequestDelegate next, IResponseCache cache)
         public async Task Invoke(HttpContext context)
         {
             var cachingContext = new ResponseCachingContext(context, _cache);
+
             // Should we attempt any caching logic?
-            if (cachingContext.CheckRequestAllowsCaching())
+            if (cachingContext.RequestIsCacheable())
             {
                 // Can this request be served from cache?
                 if (await cachingContext.TryServeFromCacheAsync())
                 {
                     return;
                 }
 
-                // Hook up to listen to the response stream
-                cachingContext.HookResponseStream();
-
                 try
                 {
+                    // Hook up to listen to the response stream
+                    cachingContext.HookResponseStream();
+
+                    // Subscribe to OnStarting event
+                    context.Response.OnStarting(OnStartingCallback, cachingContext);
+
                     await _next(context);
 
                     // If there was no response body, check the response headers now. We can cache things like redirects.
-                    if (!cachingContext.ResponseStarted)
-                    {
-                        cachingContext.OnResponseStarting();
-                    }
+                    cachingContext.OnResponseStarting();
+
                     // Finalize the cache entry
-                    cachingContext.FinalizeCaching();
+                    cachingContext.FinalizeCachingBody();
                 }
                 finally
```

Comment:
No need, response caching is only completed if everything succeeds. Early exit by default behaves as if caching was disabled.